zapcore/write_syncer.go

```diff
@@ -75,6 +77,114 @@ func (s *lockedWriteSyncer) Sync() error {
 	return err
 }
 
+type bufferWriterSyncer struct {
+	sync.Mutex
+
+	stop         chan struct{}
+	bufferWriter *bufio.Writer
+	ticker       *time.Ticker
+}
+
+const (
+	// _defaultBufferSize specifies the default size used by Buffer.
+	_defaultBufferSize = 256 * 1024 // 256 kB
+
+	// _defaultFlushInterval specifies the default flush interval for
+	// Buffer.
+	_defaultFlushInterval = 30 * time.Second
+)
+
+// Buffer wraps a WriteSyncer to buffer its output. The returned WriteSyncer
+// flushes its output as the buffer fills up, or at the provided interval,
+// whichever comes first.
+//
+// Call the returned function to finish using the WriteSyncer and flush
+// remaining bytes.
+//
+//   func main() {
+//     // ...
+//     ws, closeWS := zapcore.Buffer(ws, 0, 0)
+//     defer closeWS()
+//     // ...
+//   }
+//
+// The buffer size defaults to 256 kB if set to zero.
+// The flush interval defaults to 30 seconds if set to zero.
+func Buffer(ws WriteSyncer, bufferSize int, flushInterval time.Duration) (_ WriteSyncer, close func() error) {
+	if _, ok := ws.(*bufferWriterSyncer); ok {
+		// no need to layer on another buffer
+		return ws, func() error { return nil }
+	}
+
+	if bufferSize == 0 {
+		bufferSize = _defaultBufferSize
+	}
+
+	if flushInterval == 0 {
+		flushInterval = _defaultFlushInterval
+	}
+
+	bws := &bufferWriterSyncer{
+		stop:         make(chan struct{}),
+		bufferWriter: bufio.NewWriterSize(ws, bufferSize),
+		ticker:       time.NewTicker(flushInterval),
+	}
+
+	go bws.flushLoop()
+
+	return bws, bws.close
+}
+
+// Write writes log data into buffer syncer directly, multiple Write calls will be batched,
+// and log data will be flushed to disk when the buffer is full or periodically.
+func (s *bufferWriterSyncer) Write(bs []byte) (int, error) {
+	s.Lock()
+	defer s.Unlock()
+
+	// To avoid partial writes from being flushed, we manually flush the existing buffer if:
+	// * The current write doesn't fit into the buffer fully, and
+	// * The buffer is not empty (since bufio will not split large writes when the buffer is empty)
+	if len(bs) > s.bufferWriter.Available() && s.bufferWriter.Buffered() > 0 {
```

Comment:
this condition is not tested in the tests. even if I comment out this block, the tests don't notice any issues,
```
> git diff
diff --git a/zapcore/write_syncer.go b/zapcore/write_syncer.go
index dcb0247..844c109 100644
--- a/zapcore/write_syncer.go
+++ b/zapcore/write_syncer.go
@@ -144,11 +144,11 @@ func (s *bufferWriterSyncer) Write(bs []byte) (int, error) {
        // To avoid partial writes from being flushed, we manually flush the existing buffer if:
        // * The current write doesn't fit into the buffer fully, and
        // * The buffer is not empty (since bufio will not split large writes when the buffer is empty)
-       if len(bs) > s.bufferWriter.Available() && s.bufferWriter.Buffered() > 0 {
-               if err := s.bufferWriter.Flush(); err != nil {
-                       return 0, err
-               }
-       }
+       //if len(bs) > s.bufferWriter.Available() && s.bufferWriter.Buffered() > 0 && false {
+       //      if err := s.bufferWriter.Flush(); err != nil {
+       //              return 0, err
+       //      }
+       //}

        return s.bufferWriter.Write(bs)
 }
> go test
invalid value "unknown" for flag -level: unrecognized level: "unknown"
Usage of test:
  -level value
        minimum enabled logging level
PASS
ok      go.uber.org/zap 1.007s
```